README.md
# EDMC-Autopath
EDMC-Autopath is a plugin for [Elite: Dangerous Market Connector](https://github.com/Marginal/EDMarketConnector). This plugin provides a basic GUI to plan routes using the [neutron star plotter](https://www.spansh.co.uk/plotter/). If enabled the plugin will automatically update your system clipboard to the next system in route. [See it in action](https://www.youtube.com/watch?v=UDEgrHiYK24).

## Installation
1. Install [Elite: Dangerous Market Connector](https://github.com/Marginal/EDMarketConnector)(tested on 3.41+, might work on older versions but no guarantees).
2. Create new folder (recommended name: `autopath`) in EDMC plugin folder.
- You can open said location by using __File -> Settings -> Plugins -> Open__ in EDMC.
- On Windows, your folder will most likely be __{DRIVE}:\Users\\{username}\AppData\Local\EDMarketConnector\plugins__ (where {username} is your - Windows' username without the brackets and {DRIVE}: is drive you system is installed on).
- If you create folder with recommended name you will put files in step 3 into __{DRIVE}:\Users\\{username}\AppData\Local\EDMarketConnector\plugins\autopath__ folder 
3. Download all files in this repository (including the `pyperclip` folder) and copy them to the created folder (default: <edmc plugin folder>/autopath/).

## Ideas & Troubleshooting
If you have any ideas on how to improve this plugin, create a pull request with your changes or describe what you would like in an [issue](https://github.com/Sognus/EDMC-Autopath/issues). If you find any bugs, please [create an issue](https://github.com/Sognus/EDMC-Autopath/issues).


CHANGELOG.md
### Version 1.1.1
- Fixed an issue that caused GUI to remain expanded after disabling all features
- Fixed an issue that caused GUI not to render at all with enabled features
- Changed a default setting of features. Now all of them are being enabled on first launch. (Note: If you are updating from older version, features will change to enabled)
- Just quick hotfix - no new features for now

### Version 1.1.0
- GUI handling completely rewritten
- Added information about calculated route's total and completed systems.
- Changed label text for enabling and disabling automated copying to clipboard
- Fixed an issue that prevented ED Recon and possibly other plugins from working with EDMC Autopath enabled 

### Version 1.0.0
- Plugin released

gui.py
import ttk
import Tkinter as tk
from config import config
import myNotebook as nb
from ttkHyperlinkLabel import HyperlinkLabel
import types
from neutron import Neutron
from riches import Riches
from blank import Blank


class MainGUI(tk.Frame):
    """
    For new page:
        Set new ID for that page
        Create setup function to create menu and page
        You can see how its done in setup_neutron for example
    """

    def __init__(self, master, global_data, **kw):
        self.BLANK_ID = 0
        self.NEUTRON_ID = 1
        self.RICHES_ID = 2

        tk.Frame.__init__(self, master, **kw)
        self.main_frame = None
        self.menu_frame = None
        self.globals = global_data
        self.pages = dict()
        self.menu_buttons = dict()
        self.page_active = 0
        self.setup()
        self.update()

    def setup(self):
        """
        Creates main frame and menu frame
        :return:
        """
        # Plugin's frame
        self.globals.logger.debug("Creating GUI's main frame...")
        self.main_frame = tk.Frame(self, borderwidth=2, relief="flat")
        self.main_frame.pack(side="top", fill="both", expand=True, padx=0, pady=0)
        self.globals.logger.debug("GUI's main frame created")

        # Plugin's menu frame
        self.globals.logger.debug("Creating GUI's menu frame..")
        self.menu_frame = tk.Frame(self.main_frame, borderwidth=0, highlightthickness=0)
        self.setup_riches()
        self.setup_neutron()
        self.setup_blank()
        self.menu_frame.pack(side="top", fill="both", expand=True, anchor="w")
        self.globals.logger.debug("GUI's menu frame created.")

    def setup_blank(self):
        """
        Creates blank GUI page
        """
        
        self.globals.logger.debug("Creating GUI blank frame..")
        ID = self.BLANK_ID
        
        # MENU
        # None
        
        # PAGE
        self.pages[ID] = Blank(self.main_frame, self.globals)
        
        self.globals.logger.debug("GUI blank frame created.")

    def setup_neutron(self):
        """
        Creates Neutron Plotter GUI
        """

        self.globals.logger.debug("Creating neutron plotter frame GUI...")

        # MENU
        ID = self.NEUTRON_ID
        self.menu_buttons[ID] = tk.Button(self.menu_frame, text="Neutron", relief=tk.FLAT,
                                                   command=lambda page=ID: self.page(page))

        # PAGE
        self.pages[ID] = Neutron(self.main_frame, self.globals)

        self.globals.logger.debug("Neutron plotter frame GUI created.")

    def setup_riches(self):
        """
        Creates Road to Riches GUI
        :return:
        """

        self.globals.logger.debug("Creating road to riches frame GUI...")

        # MENU
        ID = self.RICHES_ID
        self.menu_buttons[ID] = tk.Button(self.menu_frame, text="Riches", relief=tk.FLAT,
                                                   command=lambda page=ID: self.page(page))

        # PAGE
        self.pages[ID] = Riches(self.main_frame, self.globals)

        self.globals.logger.debug("Neutron road to riches GUI created.")

    def update(self):
        self.globals.logger.debug("Updating GUI")

        # Are pages enabled
        neutron = config.getint("autopath_neutron") # page 1
        riches = config.getint("autopath_riches")   # page 2

        # Menu column
        column = 0
        button_len = 0

        # Hide all buttons
        self.globals.logger.debug("Hiding all menu buttons")
        for button in self.menu_buttons.values():
            button.grid_forget()
        self.globals.logger.debug("Menu buttons hidden.")


        # Neutron button shown
        if neutron:
            button_len = button_len + 1
            self.globals.logger.debug("Neutron plotter menu button shown.")
            self.menu_buttons[self.NEUTRON_ID].grid(row=0, column=column)
            column = column + 1
            if self.page_active == 0:
                self.page_active = self.NEUTRON_ID
        else:
            if self.page_active == self.NEUTRON_ID:
                self.page_active = 0

        # Road to riches button shown
        # Todo: Road to riches unimplememted so hardcode disabled
        if riches and False:
            button_len = button_len + 1
            self.globals.logger.debug("Road to riches menu button shown")
            self.menu_buttons[self.RICHES_ID].grid(row=0, column=column)
            column = column + 1
            if self.page_active == 0:
                self.page_active = self.RICHES_ID
        else:
            if self.page_active == self.RICHES_ID:
                self.page_active = 0

        self.menu_frame.pack_forget()
        if button_len > 2:
            self.menu_frame.pack(side="top", fill="both", expand=True, anchor="w")

        # Hide all pages
        self.globals.logger.debug("Hide all pages")
        for page in self.pages.values():
            page.pack_forget()
        self.globals.logger.debug("All pages hidden.")

        # Show page
        self.globals.logger.debug("Trying to show page number " + str(self.page_active))
        if self.page_active in self.pages.keys():
            self.globals.logger.debug("Page " + str(self.page_active) + " exist, enabled")
            self.pages[self.page_active].pack(side="top", fill="both", expand=True, anchor="w")

    def page(self, page):
        self.globals.logger.debug("Changing GUI page to " + str(page))
        self.page_active = page
        self.update()


class PrefGUI:
    # Settings for pref layout
    PADX = 10
    BUTTONX = 12

    # Storage for checkboxes
    neutron = tk.IntVar(value=config.getint('autopath_neutron'))
    riches = tk.IntVar(value=config.getint('autopath_riches'))

    @staticmethod
    def create_gui(parent, update_func, globals):
        """
        Create plugin's preference GUI
        :param parent: parent frame
        :param gui:
        :return: plugin's preference GUI
        """

        # Settings frame
        frame = nb.Frame(parent)
        frame.columnconfigure(1, weight=1)

        # Headline label
        globals.logger.debug("Creating settings hyperling label..")
        HyperlinkLabel(frame, text='EDMC: Autopath', background=nb.Label().cget('background'),
                       url='https://github.com/Sognus/EDMC-Autopath', underline=True).grid(columnspan=2,
                                                                                           padx=PrefGUI.PADX,
                                                                                           sticky=tk.W)

        globals.logger.debug("Settings hyperlink label created.")

        # Checkbox for Road To Riches
        globals.logger.debug("Creating settings neutron plotter checkbox...")
        neutron_button = nb.Checkbutton(frame, text=_('Enable neutron plotter feature'), variable=PrefGUI.neutron,
                                        command=lambda update_func=update_func: PrefGUI.prefs_changed(update_func, globals))
        neutron_button.grid(columnspan=2, padx=PrefGUI.BUTTONX, pady=(5, 0), sticky=tk.W)
        globals.logger.debug("Settings neutron plotter created.")

        # Checkbox for Road To Riches
        globals.logger.debug("Creating settings road to riches checkbox")
        riches_button = nb.Checkbutton(frame, text=_('Enable road to riches feature'), variable=PrefGUI.riches,
                                       command=lambda update_func=update_func: PrefGUI.prefs_changed(update_func, globals))
        riches_button.grid(columnspan=2, padx=PrefGUI.BUTTONX, pady=(5, 0), sticky=tk.W)
        globals.logger.debug("Settings neutron plotter created")

        return frame

    @staticmethod
    def prefs_changed(update_func=None, globals=None):
        """
        Save changes to config storage (whatever is used - windows registers or file)
        If update func is set, func will be called
        :return: None
        """
        globals.logger.debug("Saving settings..")
        config.set("autopath_neutron", PrefGUI.neutron.get())
        config.set("autopath_riches",  PrefGUI.riches.get())
        globals.logger.debug("Settings saved.")
        
        globals.logger.debug("Trying to invoke GUI update function.")
        
        if update_func is not None and isinstance(update_func, types.FunctionType):
            globals.logger.debug("Calling GUI update function...")
            update_func()
            globals.logger.debug("GUI update function completed")
        else:
            globals.logger.debug("GUI update function cannot be invoked")


load.py
from gui import *
import sys
import logging
import os
from config import config

# Plugin name
PLUGIN_NAME = 'Autopath'
DEBUG = False
ROLL_LOGS = False

# For holding module globals
globals = sys.modules[__name__]
globals.gui = None


def gui_update():
    if globals.gui is not None:
        globals.logger.debug("Updating GUI")
        globals.gui.update()


def setup_logger():
    # Logging
    globals.dir_path = os.path.dirname(os.path.realpath(__file__))
    globals.logger = logging.getLogger("edmc-autopath")
    globals.logger.setLevel(logging.DEBUG)
    globals.log_file = os.path.join(globals.dir_path, 'edmc-autopath.log')

    # Roll log files
    if os.path.exists(globals.log_file) and ROLL_LOGS:
        i = 1
        while os.path.exists(os.path.join(globals.dir_path, os.path.basename(globals.log_file) + "." + str(i) + ".log")):
            i = i + 1
        os.rename(globals.log_file, os.path.join(globals.dir_path, os.path.basename(globals.log_file) + "." + str(i) + ".log"))

    fh = logging.FileHandler(globals.log_file, "a")
    fh.setLevel(logging.DEBUG)
    formatter = logging.Formatter("%(asctime)s: %(message)s", "%d.%m.%Y %H:%M:%S")
    fh.setFormatter(formatter)
    globals.logger.addHandler(fh)

    # Enable
    globals.logger.disabled = not DEBUG



def plugin_start(plugin_dir):
    """
    Load this plugin into EDMC
    """

    # Logging
    setup_logger()

    globals.current_system = ""
    globals.last_system = ""
    
    # Checking existence of configuration values
    first_run_done = config.getint("autopath_first_run_done")
    if first_run_done == 0:
        globals.logger.info("First run of application detected")
        config.set("autopath_neutron", 1)
        config.set("autopath_riches",  1)
        config.set("autopath_first_run_done", 1)
    
    print('{}: PLUGIN STARTED'.format(PLUGIN_NAME))
    globals.logger.info('{}: PLUGIN STARTED'.format(PLUGIN_NAME))


def plugin_stop():
    """
    EDMC is closing
    """
    globals.logger.info('{}: PLUGIN DISABLED'.format(PLUGIN_NAME))
    globals.logger.info('*************************************************************')
    print('{}: PLUGIN DISABLED'.format(PLUGIN_NAME))

def plugin_app(parent):
    """
    Creates plugin's GUI frame and returns it
    :param parent: EDMC parent frame
    :return: created frame
    """
    globals.logger.debug("Creating GUI...")
    globals.gui = MainGUI(parent, globals)
    return globals.gui


def plugin_prefs(parent, cmdr, is_beta):
   """
   Return a TK Frame for adding to the EDMC settings dialog.
   """
   globals.logger.debug("Creating settings GUI...")
   return PrefGUI.create_gui(parent, gui_update, globals)


def dashboard_entry(cmdr, is_beta, entry):
    globals.logger.debug("Dashboard entry received")
    pass


def cmdr_data(data, is_beta):
    globals.logger.debug("CMDR DATA received")

    # Update current system
    if "lastSystem" in data.keys():
        dct = data["lastSystem"]
        if "name" in dct.keys():
            new = dct["name"]
            if len(globals.current_system) < 1:
                globals.current_system = new
            else:
                globals.last_system = globals.current_system
                globals.current_system = new
                globals.neutron.update_clipboard(globals.current_system)


def journal_entry(cmdr, is_beta, system, station, entry, state):
    globals.logger.debug("Journal entry received")

    # Update current system
    if entry is not None and entry["event"].lower() in ("fsdjump","startup"):
        if len(globals.current_system) < 1:
            globals.current_system = entry["StarSystem"]
        else:
            globals.last_system = globals.current_system
            globals.current_system = entry["StarSystem"]
            globals.neutron.update_clipboard(globals.current_system)


riches.py
import Tkinter as tk
import pyperclip


class Riches(tk.Frame):

    def __init__(self, master, global_data, **kw):
        tk.Frame.__init__(self, master, **kw)
        self.globals = global_data
        self.globals.neutron = self
        self.plotter = None
        self.setup()

    def setup(self):
        # Unified width
        width = 16

        self.globals.logger.debug("Riches -> Creating road to riches GUI")

        # Label origin
        self.label_origin = tk.Label(self, text="Origin: ", justify=tk.LEFT)
        self.label_origin.grid(row=0, column=0, sticky=tk.W, pady=(10, 0))

        self.globals.logger.debug("Riches -> road riches gui label origin done...")

        # Entry origin
        self.entry_origin = tk.Entry(self, width=width)
        self.entry_origin.grid(row=0, column=1, sticky=tk.E, pady=(10, 0))

        self.globals.logger.debug("Riches -> road riches gui entry origin done...")

        # Label destination
        self.label_destination = tk.Label(self, text="Destination: ", justify=tk.LEFT)
        self.label_destination.grid(row=1, column=0, sticky=tk.W)

        self.globals.logger.debug("Riches -> road riches gui label destination done...")

        # Entry destination
        self.entry_destination = tk.Entry(self, width=width)
        self.entry_destination.grid(row=1, column=1, sticky=tk.E)

        self.globals.logger.debug("Riches -> road riches gui entry destination done...")

        self.columnconfigure(6, weight=1)

        self.globals.logger.debug("Riches -> Road to riches GUI created")



plotter.py
import json
import time
from threading import Lock
from threading import Thread
import requests

# Core API link
API = "https://spansh.co.uk/api/"
# Timeout value
TIMEOUT = 30


class Plotter:

    def __init__(self):
        self.origin = None
        self.destination = None
        self.efficiency = None
        self.range = None
        # Special data handling
        self.route = None
        self.job = None
        # synchronize
        self.lock = Lock()

    def request_calculation(self, origin, destination, efficiency, ship_range):
        self.origin = origin
        self.destination = destination
        self.efficiency = efficiency
        self.range = ship_range

        # Prepare link for job queue
        link_job = API + "route?from={}&to={}&efficiency={}&range={}".format(origin, destination,efficiency, ship_range)
        # API calls for job queue
        request = requests.get(link_job).json()
        self.job = request["job"]

        thread = Thread(target=self.wait_for_route)
        thread.start()
        # After timeout we dont care anymore
        thread.join(timeout=TIMEOUT+1)

        # synchronize critical path
        with self.lock:
            return self.route

    def wait_for_route(self,):
        if self.job is None or len(self.job) < 1:
            return
        # Prepare link for route data
        link_data = API + "results/{}".format(self.job)
        # WAIT for data
        data = None
        time_start = time.time()
        data_request = requests.get(link_data)
        while data_request.status_code != 200:
            data_request = requests.get(link_data)
            time_elapsed = time.time() - time_start
            if time_elapsed > TIMEOUT:
                break

        # Get json
        data = data_request.json()
        # Synchronize
        with self.lock:
            self.route = data




autopath.py
from Tkinter import DoubleVar
from plotter import Plotter
import pyperclip
import Tkinter as tk
import ttk


class Autopath:

    def __init__(self, pointer):
        # Pre-Init
        self.pointer = pointer
        self.plotter = pointer.plotter
        self.frame = None
        self.label_origin = None
        self.separator = None
        self.entry_origin = None
        self.entry_destination = None
        self.label_efficiency = None
        self.entry_efficiency = None
        self.button_calculate = None
        self.label_range = None
        self.entry_range = None
        self.route = list()
        self.route_ready = False
        self.label_status = None
        self.checkbox_clipboard = None
        self.checkbox_status = None
        pass

    def update_clipboard(self, arrived_to):
        # Check none
        if self.checkbox_status is not None or self.checkbox_clipboard is None:
            # Check for route ready and checkbox
            if self.route_ready and self.checkbox_status.get() == 1:
                try:
                    index = self.route.index(arrived_to)
                    pyperclip.copy(self.route[index+1])
                    self.update_status("Clipboard updated to: ")
                    self.status_append(self.route[index+1])
                except:
                    # Ignore errors
                    pass

    def calculate_path(self):
        origin = self.entry_origin.get()
        dest = self.entry_destination.get()
        eff = -1
        range = -1.0

        self.update_status("")

        try:
            eff = int(self.entry_efficiency.get())
            range = float(self.entry_range.get())
        except ValueError:
            pass

        # If origin is empty, use current system
        if len(origin) < 1:
            origin = self.pointer.current_system

        # Check for data validation
        if len(origin) < 1:
            self.status_append("ERROR: Origin must not be empty!")
        if len(dest) < 1:
            self.status_append("ERROR: Destination must not be empty!")
        if eff < 1 or eff > 100:
            self.status_append("ERROR: Efficiency must be number between 1 and 100!")
        if range < 0:
            self.status_append("ERROR: Range must be decimal and not empty!")

        # Check for errors
        if len(self.label_status["text"]) > 0:
            self.route_ready = False
            return False

        self.update_status("Calculating..")

        # request calculation
        route = self.plotter.request_calculation(origin, dest, eff, range)

        self.update_status("Processing data..")

        # Process json for system names
        self.route = list()
        for record in route["result"]["system_jumps"]:
            self.route.append(record["system"])

        # Mark as ready
        if len(self.route) > 0:
            self.update_status("Route calculated..")
            self.route_ready = True
            # simulate arrive to origin system
            self.update_clipboard(self.route[0])
        else:
            self.update_status("ERROR: Route has less than one system")
            self.route_ready = False

    def status_append(self, status_text):
        if self.label_status is not None:
            self.label_status["text"] = self.label_status["text"] + "\n" + status_text
            self.label_status.update()

    def update_status(self, status_text):
        if self.label_status is not None:
            self.label_status["text"] = status_text
            self.label_status.update()

    def create_gui(self, parrent):
        # Unified width
        width = 16

        # Frame
        self.frame = tk.Frame(parrent)

        # Label origin
        self.label_origin = tk.Label(self.frame, text="Origin: ", justify=tk.LEFT)
        self.label_origin.grid(row=0, column=0, sticky=tk.W, pady=(10, 0))

        # Entry origin
        self.entry_origin = tk.Entry(self.frame, width=width)
        self.entry_origin.grid(row=0, column=1, sticky=tk.E, pady=(10, 0))
        self.entry_origin.update()

        # Label origin
        self.label_origin = tk.Label(self.frame, text="Destination: ", justify=tk.LEFT)
        self.label_origin.grid(row=1, column=0, sticky=tk.W, )

        # Entry destination
        self.entry_destination = tk.Entry(self.frame, width=width)
        self.entry_destination.grid(row=1, column=1, sticky=tk.E)
        self.entry_destination.update()

        # Label range
        self.label_range = tk.Label(self.frame, text="Range (LY): ", justify=tk.LEFT)
        self.label_range.grid(row=2, column=0, sticky=tk.W)

        # Entry range
        range_default = tk.DoubleVar(value=50.0)
        self.entry_range = tk.Entry(self.frame, width=width, textvariable=range_default)
        self.entry_range.grid(row=2, column=1, sticky=tk.E)
        self.entry_range.update()

        # Label efficiency
        self.label_efficiency = tk.Label(self.frame, text="Efficiency (%): ", justify=tk.LEFT)
        self.label_efficiency.grid(row=3, column=0, sticky=tk.W)

        # Efficiency entry
        efficiency_default = tk.IntVar(value=60)
        self.entry_efficiency = tk.Spinbox(self.frame, from_=1, to=100, width=width - 2, textvariable=efficiency_default)
        self.entry_efficiency.grid(row=3, column=1, sticky=tk.E)

        # Checkbox clipboard
        self.checkbox_status = tk.IntVar(value=1)
        self.checkbox_clipboard = tk.Checkbutton(self.frame, text="Auto-copy to clipboard", variable=self.checkbox_status)
        self.checkbox_clipboard.grid(row=4, column=0, columnspan=2, sticky=tk.W)

        # Button
        self.button_calculate = tk.Button(self.frame, text="Calculate", command=self.calculate_path)
        self.button_calculate.grid(row=5, column=0, columnspan=2, pady=(3, 0))

        # Label Status
        self.label_status = tk.Label(self.frame, text="Ready.. ", justify=tk.CENTER)
        self.label_status.grid(row=6, column=0, columnspan=2, pady=(0,10))

        # Frame layout
        self.frame.columnconfigure(6, weight=1)
        
        return self.frame


neutronplotter.py
import json
import time
from threading import Lock
from threading import Thread
import requests

# Core API link
API = "https://spansh.co.uk/api/"
# Timeout value
TIMEOUT = 30


class NeutronPlotter:

    def __init__(self, globals):
        self.globals = globals
        self.origin = None
        self.destination = None
        self.efficiency = None
        self.range = None
        # Special data handling
        self.route = None
        self.job = None
        # synchronize
        self.lock = Lock()

    def request_calculation(self, origin, destination, efficiency, ship_range):
        self.globals.logger.debug("Neutron plotter -> calculation requested")
        self.origin = origin
        self.destination = destination
        self.efficiency = efficiency
        self.range = ship_range

        # Prepare link for job queue
        link_job = API + "route?from={}&to={}&efficiency={}&range={}".format(origin, destination,efficiency, ship_range)
        self.globals.logger.debug("Neutron -> request target is {}".format(link_job))
        # API calls for job queue
        request = requests.get(link_job).json()
        self.job = request["job"]

        thread = Thread(target=self.wait_for_route)
        thread.start()
        # After timeout we dont care anymore
        thread.join(timeout=TIMEOUT+1)

        # synchronize critical path
        with self.lock:
            return self.route

    def wait_for_route(self,):
        if self.job is None or len(self.job) < 1:
            return
        # Prepare link for route data
        link_data = API + "results/{}".format(self.job)
        # WAIT for data
        data = None
        time_start = time.time()
        data_request = requests.get(link_data)
        while data_request.status_code != 200:
            data_request = requests.get(link_data)
            time_elapsed = time.time() - time_start
            if time_elapsed > TIMEOUT:
                break

        # Get json
        data = data_request.json()
        # Synchronize
        with self.lock:
            self.route = data




neutron.py
import Tkinter as tk
import pyperclip
from neutronplotter import NeutronPlotter
import math
from config import config
try:
    from EDMCOverlay import edmcoverlay
except ImportError:  ## test mode
    edmcoverlay = None

class Neutron(tk.Frame):

    def __init__(self, master, global_data, **kw):
        tk.Frame.__init__(self, master, **kw)
        self.overlay = None
        self.globals = global_data
        self.globals.neutron = self
        self.plotter = NeutronPlotter(self.globals)
        self.route_ready = False
        self.setup()

    def update_location(self, arrived_to):
        self.globals.logger.debug("Neutron -> Updating system clipboard")
        if self.route_ready:
            self.globals.logger.debug("Neutron -> Route is ready")
            index = self.route.index(arrived_to)
            total = len(self.route)
            percent = float(float(((index + 1)) / float(total)) * float(100))
            percent = math.floor(percent * 100) / 100.0
            self.update_status("Clipboard updated to: ")
            self.status_append(self.route[index + 1])
            self.status_append("")
            self.status_append("{}/{} ({:.2f}%)".format(index + 1, total, percent))

            if config.getint("autopath_clipboard"):
                self.globals.logger.debug("Neutron -> Route is ready and clipboard checkbox is checked")
                try:
                    pyperclip.copy(self.route[index + 1])
                    self.globals.logger.debug("Neutron -> system clipboard changed to {}".format(self.route[index + 1]))
                except Exception as e:
                    print(e)
                    self.globals.logger.debug("Neutron -> clipboard update failed > {}".format(str(e)))
                    pass
            if config.getint("autopath_overlay"):
                self.display_overlay("Waypoint reached: {}/{} ({:.2f}%)".format(index + 1, total, percent))


    def status_append(self, status_text):
        if self.label_status is not None:
            self.globals.logger.debug("Neutron -> Appening {} to status text".format(status_text))
            self.label_status["text"] = self.label_status["text"] + "\n" + status_text
            self.label_status.update()

    def update_status(self, status_text):
        if self.label_status is not None:
            self.globals.logger.debug("Neutron -> setting status text to {}".format(status_text))
            self.label_status["text"] = status_text
            self.label_status.update()

    def display_overlay(self, message, timeout_seconds=15):
        if self.overlay is None:
            self.overlay = edmcoverlay.Overlay()
        try:
            self.overlay.send_message("Neutron", message, "#aaf9ff", x=520, y=120, ttl=timeout_seconds, size="large")
            self.globals.logger.debug("Neutron -> displayed overlay message: {}".format(message))
        except Exception as err:
            self.globals.logger.debug("Neutron -> failed to connect to overlay: {}".format(err))

    def calculate_path(self):
        self.display_overlay("Calculating route...", 5)
        self.globals.logger.debug("Neutron -> calculating path")
        origin = self.entry_origin.get()
        dest = self.entry_destination.get()
        eff = -1
        range = -1.0

        self.update_status("")

        try:
            eff = int(self.entry_efficiency.get())
            range = float(self.entry_range.get())
        except ValueError:
            pass

        # If origin is empty, use current system
        if len(origin) < 1:
            self.globals.logger.debug("Neutron -> origin is not set, using current system {}".format(self.globals.current_system))
            origin = self.globals.current_system

        # Check for data validation
        if len(origin) < 1:
            self.globals.logger.debug("Neutron -> Origin cannot be empty (error)")
            self.status_append("ERROR: Origin must not be empty!")
        if len(dest) < 1:
            self.globals.logger.debug("Neutron -> Destination cannot be empty (error)")
            self.status_append("ERROR: Destination must not be empty!")
        if eff < 1 or eff > 100:
            self.globals.logger.debug("Neutron -> Efficiency must be number between 1 and 100! (error)")
            self.status_append("ERROR: Efficiency must be number between 1 and 100!")
        if range < 0:
            self.globals.logger.debug("Neutron -> Range must be decimal and not empty! (error)")
            self.status_append("ERROR: Range must be decimal and not empty!")

        # Check for errors
        if len(self.label_status["text"]) > 0:
            self.globals.logger.debug("Neutron -> Errors were detected, calculation will not start (error)")
            self.route_ready = False
            return False

        self.update_status("Calculating..")

        # request calculation
        self.globals.logger.debug("Neutron -> Requesting route..")
        route = self.plotter.request_calculation(origin, dest, eff, range)

        self.update_status("Processing data..")

        # Process json for system names
        self.globals.logger.debug("Neutron -> Processing data")
        self.route = list()
        for record in route["result"]["system_jumps"]:
            self.globals.logger.debug("Neutron -> system found {}".format(record["system"]))
            self.route.append(record["system"])


        # Mark as ready
        if len(self.route) > 0:
            self.globals.logger.debug("Neutron -> Route is ready")
            self.update_status("Route calculated..")
            self.route_ready = True
            # simulate arrive to origin system
            self.update_location(self.route[0])
        else:
            self.globals.logger.debug("Neutron -> Route is less than one system away = destination is origin (error)")
            self.update_status("ERROR: Route has less than one system")
            self.route_ready = False

    def setup(self):
        # Unified width
        width = 16

        self.globals.logger.debug("Neutron -> Creating neutron plotter GUI")

        # Label origin
        self.label_origin = tk.Label(self, text="Origin: ", justify=tk.LEFT)
        self.label_origin.grid(row=0, column=0, sticky=tk.W, pady=(10, 0))

        # Entry origin
        self.entry_origin = tk.Entry(self, width=width)
        self.entry_origin.grid(row=0, column=1, sticky=tk.E, pady=(10, 0))

        # Label destination
        self.label_destination = tk.Label(self, text="Destination: ", justify=tk.LEFT)
        self.label_destination.grid(row=1, column=0, sticky=tk.W)

        # Entry destination
        self.entry_destination = tk.Entry(self, width=width)
        self.entry_destination.grid(row=1, column=1, sticky=tk.E)

        # Label range
        self.label_range = tk.Label(self, text="Range (LY): ", justify=tk.LEFT)
        self.label_range.grid(row=2, column=0, sticky=tk.W)

        # Entry range
        range_default = tk.DoubleVar(value=50.0)
        self.entry_range = tk.Entry(self, width=width, textvariable=range_default)
        self.entry_range.grid(row=2, column=1, sticky=tk.E)

        # Label efficiency
        self.label_efficiency = tk.Label(self, text="Efficiency (%): ", justify=tk.LEFT)
        self.label_efficiency.grid(row=3, column=0, sticky=tk.W)

        # Efficiency entry
        efficiency_default = tk.IntVar(value=60)
        self.entry_efficiency = tk.Spinbox(self, from_=1, to=100, width=width - 2,
                                           textvariable=efficiency_default)
        self.entry_efficiency.grid(row=3, column=1, sticky=tk.E)

        # Button
        self.button_calculate = tk.Button(self, text="Calculate", command=self.calculate_path)
        self.button_calculate.grid(row=4, column=0, columnspan=2, pady=(3, 0))

        # Label Status
        self.label_status = tk.Label(self, text="Ready.. ", justify=tk.CENTER)
        self.label_status.grid(row=5, column=0, columnspan=2, pady=(0, 10))

        # Frame layout
        self.columnconfigure(6, weight=1)

        self.globals.logger.debug("Neutron -> neutron plotter GUI completed")